creator/text_chunk.go

```diff
@@ -26,17 +31,112 @@ type TextChunk struct {
 	annotationProcessed bool
 }
 
+// NewTextChunk returns a new text chunk instance.
+func NewTextChunk(text string, style TextStyle) *TextChunk {
+	return &TextChunk{
+		Text:  text,
+		Style: style,
+	}
+}
+
 // SetAnnotation sets a annotation on a TextChunk.
 func (tc *TextChunk) SetAnnotation(annotation *model.PdfAnnotation) {
 	tc.annotation = annotation
 }
 
-// newTextChunk returns a new text chunk instance.
-func newTextChunk(text string, style TextStyle) *TextChunk {
-	return &TextChunk{
-		Text:  text,
-		Style: style,
+// Wrap wraps the text of the chunk into lines based on its style and the
+// specified width.
+func (tc *TextChunk) Wrap(width float64) ([]string, error) {
+	if int(width) <= 0 {
+		return []string{tc.Text}, nil
 	}
+
+	var lines []string
+	var line []rune
+	var lineWidth float64
+	var widths []float64
+
+	style := tc.Style
+	runes := []rune(tc.Text)
+
+	for _, r := range runes {
+		// Move to the next line due to newline wrapping (LF).
+		if r == '\u000A' {
+			lines = append(lines, strings.TrimRightFunc(string(line), unicode.IsSpace)+string(r))
+			line = nil
+			lineWidth = 0
+			widths = nil
+			continue
+		}
+
+		metrics, found := style.Font.GetRuneMetrics(r)
+		if !found {
+			common.Log.Debug("ERROR: Rune char metrics not found! rune=0x%04x=%c font=%s %#q",
+				r, r, style.Font.BaseFont(), style.Font.Subtype())
+			common.Log.Trace("Font: %#v", style.Font)
+			common.Log.Trace("Encoder: %#v", style.Font.Encoder())
+			return nil, errors.New("glyph char metrics missing")
+		}
+
+		w := style.FontSize * metrics.Wx
+		charWidth := w + style.CharSpacing*1000.0
+		if lineWidth+w > width*1000.0 {
+			// Goes out of bounds. Break on the character.
+			idx := -1
+			for i := len(line) - 1; i >= 0; i-- {
+				if line[i] == ' ' {
+					idx = i
+					break
+				}
+			}
+			if idx > 0 {
+				// Back up to last space.
+				lines = append(lines, strings.TrimRightFunc(string(line[0:idx+1]), unicode.IsSpace))
+
+				// Remainder of line.
+				line = append(line[idx+1:], r)
+				widths = append(widths[idx+1:], charWidth)
+
+				lineWidth = 0
+				for _, width := range widths {
+					lineWidth += width
+				}
+			} else {
+				lines = append(lines, strings.TrimRightFunc(string(line), unicode.IsSpace))
+				line = []rune{r}
+				widths = []float64{charWidth}
+				lineWidth = charWidth
+			}
+		} else {
+			line = append(line, r)
+			lineWidth += charWidth
+			widths = append(widths, charWidth)
+		}
+	}
+	if len(line) > 0 {
+		lines = append(lines, string(line))
+	}
+
+	return lines, nil
+}
+
+// Fit fits the chunk into the specified bounding box, cropping off the
+// remainder in a new chunk, if it exceeds the specified dimensions.
+func (tc *TextChunk) Fit(width, height float64) (*TextChunk, error) {
+	lines, err := tc.Wrap(width)
+	if err != nil {
+		return nil, err
+	}
+
+	fit := int(height / tc.Style.FontSize)
```

Comment:
Need to account for line heigth?  Cap height?

Reply:
Follow up: It was discussed that the line height indeed needs to be accounted for.  However the line height is not a property of the Chunk but rather the Paragraph that contains the Chunk.  The problem can be averted by dividing out the line height prior to calling Fit/Wrap methods (if line height is not 1.0 - default).
In the future we may consider making the LineHeight a property of Chunk and simply using the LineHeight of Paragraph elements as a wrapper to set the lineheight of Chunks.  Needs some investigation (a ticket will be created for this).

Reply:
Issue https://github.com/unidoc/unipdf/issues/183 tracks this